runtime/src/master.rs

```
@@ -0,0 +1,695 @@
+        ///
+        /// After a sucessful execution, the current round of votes is cleared and round number is
+        /// increased.
+        #[weight = 10_000 + proposal.get_dispatch_info().weight + 1_000 * auth.len() as u64]
+        pub fn execute(
+            origin,
+            proposal: Box<<T as Trait>::Call>,
+            auth: PMAuth,
+        ) -> DispatchResult {
+            Module::<T>::execute_(origin, proposal, auth)
+        }
+
+        /// Root-only. Sets the members and vote requirement for master. Increases the round number
+        /// and removes the votes for the prevous round.
+        ///
+        /// ```
+        /// # use dock_testnet_runtime::master::Membership;
+        /// # extern crate alloc;
+        /// # use alloc::collections::BTreeSet;
+        /// #
+        /// // Setting the following membership will effectively dissolve the master account.
+        /// Membership {
+        ///     members: BTreeSet::new(),
+        ///     vote_requirement: 1,
+        /// };
+        /// ```
+        ///
+        /// Setting the vote requirement to zero grants unrestricted root access to
+        /// all accounts. It is not recomended to set the vote requirement to zero on a
+        /// production chain.
+        #[weight = 10_000 + 1_000 * membership.members.len() as u64]
+        pub fn set_members(
+            origin,
+            membership: Membership,
+        ) -> DispatchResult {
+            Module::<T>::set_members_(origin, membership)
+        }
+    }
+}
+
+impl<T: Trait> Module<T> {
+    pub fn execute_(
+        origin: T::Origin,
+        proposal: Box<<T as Trait>::Call>,
+        auth: PMAuth,
+    ) -> DispatchResult {
+        ensure_signed(origin)?;
+
+        // check
+        let membership = Members::get();
+        let payload = StateChange::MasterVote(Vote {
+            proposal: proposal.encode(),
+            round_no: Round::get(),
+        })
+        .encode();
+        ensure!(
```

Nitpick: The 2 `ensure`s can be moved before serializing to fail fast.

---

done 048d8183